pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ComputeFieldStep.java

```diff
@@ -167,12 +172,66 @@ private GenericRecord computeFields(
     }
     // Add computed fields
     for (ComputeField field : fields) {
-      newRecordBuilder.set(field.getName(), field.getEvaluator().evaluate(context));
-      ;
+      newRecordBuilder.set(
+          field.getName(),
+          getAvroValue(
+              newSchema.getField(field.getName()).schema(),
+              field.getEvaluator().evaluate(context)));
     }
     return newRecordBuilder.build();
   }
 
+  private Object getAvroValue(Schema schema, Object value) {
+    if (value == null) {
+      return null;
+    }
+
+    LogicalType logicalType = getLogicalType(schema);
+    if (logicalType == null) {
+      return value;
+    }
+
+    // Avro logical type conversion: https://avro.apache.org/docs/1.8.2/spec.html#Logical+Types
+    switch (logicalType.getName()) {
+      case "date":
+        validateLogicalType(value, schema.getLogicalType(), LocalDate.class);
+        LocalDate localDate = (LocalDate) value;
+        return (int) localDate.toEpochDay();
+      case "time-millis":
+        validateLogicalType(value, schema.getLogicalType(), LocalTime.class);
+        LocalTime localTime = (LocalTime) value;
+        return (int) (localTime.toNanoOfDay() / 1000000);
+      case "timestamp-millis":
+        validateLogicalType(value, schema.getLogicalType(), OffsetDateTime.class);
+        OffsetDateTime offsetDateTime = (OffsetDateTime) value;
+        return offsetDateTime.toInstant().toEpochMilli();
+    }
+
+    throw new IllegalArgumentException(
+        String.format("Invalid logical type %s for value %s", schema.getLogicalType(), value));
+  }
+
+  private LogicalType getLogicalType(Schema schema) {
+    if (!schema.isUnion()) {
+      return schema.getLogicalType();
+    }
+
+    return schema
+        .getTypes()
+        .stream()
+        .filter(subSchema -> subSchema.getLogicalType() != null)
+        .map(Schema::getLogicalType)
```

Comment:
The filter and map can be switched. And filter is then just `Objects::nonnull`

Reply:
+1